File: Learning-Sequelize/TSM.js
// Point = {
//   id: 'A' // UNIQUE
//   x: 5
//   y: 7
// }
//
// function(originId, [Point, Point]) {
//
//   return {
//     bestPath: [Point,Point,Point],
//     bestCost: decimal,
//     worsePath: [Point,Point,Point],
//     worseCost: decimal
//   }
// }
//
// function dist() {
//
// }
const points = [
  { id: 0, x: 0, y: 0 },
  { id: 1, x: 2, y: 1 },
  { id: 2, x: 4, y: 5 }
  // {id: 3, x: 6, y: 2 },
  // {id: 4, x: 3, y: 7 },
];

function TSM(startId, points, result = { path: [], cost: 0 }) {
  const startPoint = points.find(p => p.id === startId);
  const combination = [];

  points.forEach(point => {});

  return startPoint;
}

function permuations(points) {
  if (points.length === 1) return [points];

  const result = [];

  for (let i = 0; i < points.length; i++) {
    const subPoints = permuations([
      ...points.slice(0, i),
      ...points.slice(i + 1)
    ]);
    const subResult = subPoints.map(point => [points[i], ...subPoints]);
    console.log(subResult);
    result.concat(subResult);
  }

  return result;
}

console.log(permuations([2, 3]));

function dist(p1, p2) {
  const xDiff = Math.abs(p1.x - p2.x);
  const yDiff = Math.abs(p1.y - p2.y);
  return Math.sqrt(xDiff * xDiff + yDiff * yDiff);
}

// console.log(TSM(2, points));

// const p1 = { x: 2, y: 1 };
// const p2 = { x: 4, y: 5 };
//
// console.log(dist(p1, p2));

File: Libraries/Learning-Sequelize/TSM.rb
require 'byebug'
require 'benchmark'

points = [
  { id: 0, x: 0, y: 0 },
  { id: 1, x: 3, y: 4 },
  { id: 2, x: 4, y: 5 },
  { id: 3, x: 6, y: 2 },
  { id: 4, x: 3, y: 7 },
  { id: 5, x: 12, y: 10 },
  { id: 6, x: 43, y: 22 },
  { id: 7, x: 42, y: 28 },
  { id: 8, x: 123, y: 99 },
  { id: 9, x: 55, y: 63 },
  { id: 10, x: 555, y: 603 }
];

def TSM(startId, points)
  startPoint = points.find { |point| point[:id] == startId}
  otherPoints = points.reject { |point| point[:id] == startId }

  combination = permutations(otherPoints)
  routes = []
  combination.each do |route|
    total = 0;
    current = startPoint
    route.each do |p|
      total += dist(current, p)
      current = p
    end
    routes.push({ route: [startPoint] + route , cost: total })

  end
  routes.sort_by! { |route| route[:cost]}

  return {
    bestPath: routes.first[:route],
    bestCost: routes.first[:cost],
    worsePath: routes.last[:route],
    worseCost: routes.last[:cost]
  }
end

def permutations(points)
  return [points] if points.length == 1

  result = []
  points.each_index do |i|
    subArrays = permutations(points[0...i] + points[(i + 1)..-1])
    result += subArrays.map { |subArr| [points[i]] + subArr  }
  end
  result
end

def dist(p1, p2)
  xDiff = (p1[:x] - p2[:x]).abs
  yDiff = (p1[:y] - p2[:y]).abs
  return Math.sqrt(xDiff**2 + yDiff**2);
end

# p permutations(points)
# p dist(points[0], points[1])
time = Benchmark.measure {

  p TSM(0, points)
}
puts time
